connect-inject/health_check_resource.go

```diff
@@ -0,0 +1,283 @@
+	ConsulUrl *url.URL
+	// ReconcilePeriod is the period by which reconcile gets called, default to 1 minute
+	ReconcilePeriod time.Duration
+
+	Ctx  context.Context
+	lock sync.Mutex
+}
+
+// Run is the long-running runloop for periodically running Reconcile
+// it initially starts a Reconcile phase at startup and then calls Reconcile
+// once every ReconcilePeriod time
+func (h *HealthCheckResource) Run(stopCh <-chan struct{}) {
+	// Start the background watchers
+	err := h.Reconcile()
+	if err != nil {
+		h.Log.Error("reconcile returned an error", "err", err)
+	}
+
+	reconcileTimer := time.NewTimer(h.ReconcilePeriod)
+	defer reconcileTimer.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			h.Log.Info("HealthCheckController quitting")
+			return
+
+		case <-reconcileTimer.C:
+			err = h.Reconcile()
+			if err != nil {
+				h.Log.Error("reconcile returned an error", "err", err)
+			}
+			reconcileTimer.Reset(h.ReconcilePeriod)
+		}
+	}
+}
+
+// Delete is not implemented because it is handled by the preStop phase whereby all services
+// related to the pod are deregistered which also deregisters health checks
+func (h *HealthCheckResource) Delete(string) error {
+	return nil
+}
+
+// Informer starts a sharedindex informer which watches and lists corev1.Pod objects
+// which meet the filter of labelInject
+func (h *HealthCheckResource) Informer() cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(
+		// ListWatch takes a List and Watch function which we filter based on label which was injected
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return h.KubernetesClientset.CoreV1().Pods(metav1.NamespaceAll).List(h.Ctx,
+					metav1.ListOptions{LabelSelector: labelInject})
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return h.KubernetesClientset.CoreV1().Pods(metav1.NamespaceAll).Watch(h.Ctx,
+					metav1.ListOptions{LabelSelector: labelInject})
+			},
+		},
+		&corev1.Pod{}, // the target type (Pod)
+		0,             // no resync (period of 0)
+		cache.Indexers{},
+	)
+}
+
+// Upsert processes a create or update event.
+// Two primary use cases are handled, new pods will get a new consul TTL health check
+// registered against their respective agent and service, and updates to pods will have
+// this TTL health check updated to reflect the pod status.
+func (h *HealthCheckResource) Upsert(key string, raw interface{}) error {
+	pod, ok := raw.(*corev1.Pod)
+	if !ok {
+		return fmt.Errorf("invalid pod object")
+	}
+	if !h.shouldProcess(pod) {
+		// Skip pods that are not running or have not been properly injected
+		return nil
+	}
+	err := h.reconcilePod(pod)
+	if err != nil {
+		h.Log.Error("unable to update pod", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (h *HealthCheckResource) reconcilePod(pod *corev1.Pod) error {
+	h.Log.Debug("processing pod", "name", pod.Name)
+	if pod.Status.Phase != corev1.PodRunning {
```

Comment:
I think this should call `shouldProcess` so that both `Reconcile` and `Upsert` are "processing" the same pods. Right now `Upsert` ignores pods without `annotationInject` status, but `Reconcile` doesn't. This would also mean that the `shouldProcess` call from `Upsert` can be removed.